pkg/migrations/sql/idempotent/tracing-public.sql

```diff
@@ -0,0 +1,459 @@
+    INNER JOIN LATERAL -- inner join lateral enables partition elimination at execution time
+    (
+        SELECT
+            a.key,
+            a.value
+        FROM SCHEMA_TRACING.tag a
+        WHERE a.id = x.value::text::bigint
+        AND a.key = ANY(_keys) -- ANY works with partition elimination
+    ) a on (true)
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.jsonb(SCHEMA_TRACING_PUBLIC.tag_map) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.val(_tag_map SCHEMA_TRACING_PUBLIC.tag_map, _key SCHEMA_TRACING_PUBLIC.tag_k)
+RETURNS SCHEMA_TRACING_PUBLIC.tag_v
+AS $func$
+    SELECT a.value
+    FROM SCHEMA_TRACING.tag a
+    WHERE a.key = _key -- partition elimination
+    AND a.id = (_tag_map->>(SELECT id::text FROM SCHEMA_TRACING.tag_key WHERE key = _key))::bigint
+    LIMIT 1
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.val(SCHEMA_TRACING_PUBLIC.tag_map, SCHEMA_TRACING_PUBLIC.tag_k) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.val_text(_tag_map SCHEMA_TRACING_PUBLIC.tag_map, _key SCHEMA_TRACING_PUBLIC.tag_k)
+RETURNS text
+AS $func$
+    SELECT a.value#>>'{}'
+    FROM SCHEMA_TRACING.tag a
+    WHERE a.key = _key -- partition elimination
+    AND a.id = (_tag_map->>(SELECT id::text FROM SCHEMA_TRACING.tag_key WHERE key = _key))::bigint
+    LIMIT 1
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.val_text(SCHEMA_TRACING_PUBLIC.tag_map, SCHEMA_TRACING_PUBLIC.tag_k) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.get_tag_map(_tags jsonb)
+RETURNS SCHEMA_TRACING_PUBLIC.tag_map
+AS $func$
+    SELECT coalesce(jsonb_object_agg(a.key_id, a.id), '{}')::SCHEMA_TRACING_PUBLIC.tag_map
+    FROM jsonb_each(_tags) x
+    INNER JOIN LATERAL
+    (
+        SELECT a.key_id, a.id
+        FROM SCHEMA_TRACING.tag a
+        WHERE x.key = a.key
+        AND x.value = a.value
+        LIMIT 1
+    ) a on (true)
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.get_tag_map(jsonb) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.tag_key_id(_key SCHEMA_TRACING_PUBLIC.tag_k)
+RETURNS text
+AS $func$
+    SELECT k.id::text
+    FROM SCHEMA_TRACING.tag_key k
+    WHERE k.key = _key
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.tag_key_id(SCHEMA_TRACING_PUBLIC.tag_k) TO prom_reader;
+
+CREATE OR REPLACE FUNCTION SCHEMA_TRACING_PUBLIC.tag_key_ids(VARIADIC _keys SCHEMA_TRACING_PUBLIC.tag_k[])
+RETURNS text[]
+AS $func$
+    SELECT array_agg(k.id::text)
+    FROM SCHEMA_TRACING.tag_key k
+    WHERE k.key = ANY(_keys)
+$func$
+LANGUAGE SQL STABLE PARALLEL SAFE STRICT;
+GRANT EXECUTE ON FUNCTION SCHEMA_TRACING_PUBLIC.tag_key_ids(SCHEMA_TRACING_PUBLIC.tag_k[]) TO prom_reader;
```

Comment:
what is the usecase for the functions? when would the be used?

Reply:
"show me spans which have this tag(s)"

Reply:
```
SELECT                                                                                                                                                                     trace_id,                                                                                                                                                                         span_id,                                                                                                                                                                          val_text(span_tags, 'host.name') as host_name                                                                                                                                 FROM span                                                                                                                                                                         
WHERE span_tags ? tag_key_id('host.name')                                                                                                                                         LIMIT 10                                                                                                                                                                          
;
```

Reply:
```
SELECT
    trace_id,
    span_id,
    val(span_tags, 'host.name') as host_name,
    val(span_tags, 'host.ip') as host_ip
FROM span
WHERE span_tags ? tag_key_ids('host.name', 'host.ip')
LIMIT 10
;
```

Reply:
it's not yet "fast" though

Reply:
probably need operators for this, if we can find ones that make sense

Reply:
on second thought, i don't think these are necessary.

Reply:
okay. Removing these in favor of `has_tag()` and a new `#` operator

```
select trace_id, span_id, val(span_tags, 'bool5') from span where span_tags # 'bool5';
```